core/sta/ecc_self_tests.c

```diff
@@ -0,0 +1,227 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <trace.h>
+#include <tee/tee_cryp_provider.h>
+#include <tee_api_defines.h>
+#include <user_ta_header.h>
+
+#define TA_NAME		"ecc_self_test.ta"
+
+#define CMD_ECC_GEN_KEY_SELF_TESTS	0
+
+#define ECC_SELF_TEST_UUID \
+		{ 0xf34f4f3c, 0xab30, 0x4573,  \
+		{ 0x91, 0xBF, 0x3C, 0x57, 0x02, 0x4D, 0x51, 0x99 } }
+
+static TEE_Result create_ta(void)
+{
+	DMSG("create entry point for static ta \"%s\"", TA_NAME);
+	return TEE_SUCCESS;
+}
+
+static void destroy_ta(void)
+{
+	DMSG("destroy entry point for static ta \"%s\"", TA_NAME);
+}
+
+static TEE_Result open_session(uint32_t nParamTypes __unused,
+		TEE_Param pParams[4] __unused, void **ppSessionContext __unused)
+{
+	DMSG("open entry point for static ta \"%s\"", TA_NAME);
+	return TEE_SUCCESS;
+}
+
+static void close_session(void *pSessionContext __unused)
+{
+	DMSG("close entry point for static ta \"%s\"", TA_NAME);
+}
+
+static TEE_Result ecc_check_size(struct bignum *n, uint32_t key_size_bits)
+{
+	uint32_t size = crypto_ops.bignum.num_bits(n);
+
+	if ((size < key_size_bits-32) || (key_size_bits < size)) {
+		EMSG("Generate keysize = %d, expected %d", size, key_size_bits);
+		return TEE_ERROR_GENERIC;
+	}
+	return TEE_SUCCESS;
+}
+
+static TEE_Result ecc_check_equal(struct bignum *p, struct bignum *q)
+{
+	if (crypto_ops.bignum.compare(p, q) == 0) {
+		EMSG("Keys are equal");
+		return TEE_ERROR_GENERIC;
+	} else {
+		return TEE_SUCCESS;
+	}
+}
+
+static TEE_Result ecc_generate_key_curve(struct ecc_keypair *key1,
+					 struct ecc_keypair *key2,
+					 uint32_t curve)
+{
+	TEE_Result res;
+	uint32_t key_size_bits;
+
+	/* Generates 2 keys */
+	key1->curve = curve;
+	res = crypto_ops.acipher.gen_ecc_key(key1);
+	if (res != TEE_SUCCESS) {
+		if (curve == TEE_ECC_CURVE_NIST_P521 + 1) {
```

Comment:
I would change this to `if (curve == 0 || curve > TEE_ECC_CURVE_NIST_P521)` instead.